src/main/scala/com/gu/newsletterlistcleanse/EitherConverter.scala

```diff
@@ -3,18 +3,61 @@ package com.gu.newsletterlistcleanse
 import scala.annotation.tailrec
 
 object EitherConverter {
-    implicit class EitherList[E, A](le: List[Either[E, A]]){
-      def toEitherList: Either[E, List[A]] = {
-        @tailrec
-        def helper(list: List[Either[E, A]], acc: List[A]):
-        Either[E, List[A]] = list match {
-          case Nil => Right(acc)
-          case x::xs => x match {
-            case Left(e) => Left(e)
-            case Right(v) => helper(xs, acc :+ v)
+  implicit class EitherList[E, A](le: List[Either[E, A]]){
+    def toEitherList: Either[List[E], List[A]] = {
+      @tailrec
+      def helper(list: List[Either[E, A]], acc: Either[List[E], List[A]]): Either[List[E], List[A]] =
+        list match {
+          case Nil => acc
+          case x :: xs => x match {
+            case Left(e) => {
+              acc match {
+                case Left(es) => helper(xs, Left(e :: es))
+                case Right(_) => helper(xs, Left(List(e)) )
+              }
+            }
+            case Right(a) => {
+              acc match {
+                case Left(es) => helper(xs, Left(es))
+                case Right(as) => helper(xs, Right(a :: as))
+              }
+            }
+          }
+        }
+      helper(le.reverse, Right(Nil))
+    }
+
+    def toEitherListFold: Either[List[E], List[A]] = {
+      le.foldRight[Either[List[E], List[A]]](Right(Nil)) { (ea, acc) =>
+        ea match {
+          case Left(e) => {
+            acc match {
+              case Left(es) => Left(e :: es)
+              case Right(_) => Left(List(e))
+            }
+          }
+          case Right(a) => {
+            acc match {
+              case Left(es) => Left(es)
+              case Right(as) => Right(a :: as)
+            }
           }
         }
-        helper(le, Nil)
       }
     }
+
+    def toEitherListFold2: Either[List[E], List[A]] = {
+      le.foldRight[Either[List[E], List[A]]](Right(Nil)) {
+        case (Left(e), Left(es)) =>
+          Left(e :: es)
+        case (Left(e), Right(_)) =>
+          Left(List(e)) // Throw away accumulated successes and only track errors
+        case (Right(_), Left(es)) =>
+          Left(es)
+        case (Right(a), Right(as)) =>
+          Right(a :: as) // Keep accumulating successes
```

Comment:
We can condense this down to 4 lines if we don't need the explanatory comments.

Reply:
I think this implementation is the winner, much easier to read than the nested pattern match.

I think you may also want to use foldLeft, which is more performant with the scala `List`. Not that performance matters much in our context, we're still orders of magnitude faster than any network call.

Reply:
If we use `foldLeft` will we need to reverse the list first in order to keep the ordering? Not that ordering necessarily matters in this case.

Reply:
That's a good point. I guess ultimately it doesn't matter so foldRight it is!